### DISCUSSION TITLE
How to hot update index？

### DISCUSSION THREAD

**Discussion by qintian-qe** [oracle/opengrok#4363] [General]
After deleting the index corresponding to the project on the server, I also cleaned up the cache on the front-end and found that the front-end can still perform indexing. Where is the index file stored at this time? (The above question is raised because I have performed index parallel operations and need to notify the web side that the index file has been updated. Currently, I can only restart the web service to update the index. What should I do to notify the web side in real-time?)

**Comment by vladak** [comment]
I assume that "front-end can still perform indexing" means that the webapp can still perform search.

It depends on how the index is deleted and what kind of index access the webapp uses. Usually the Lucene within the webapp will open the index via memory mapping. This means that even if the files are deleted, the app is still able to access them, at least on Unix.

The recommended way to delete index data for a project is to use the REST API endpoint, see https://opengrok.docs.apiary.io/#reference/0/project-data-management. This should make sure that the objects governing the index data search within the webapp are updated.

**Reply by qintian-qe** [reply]
In order to achieve parallel index generation, I now incrementally generate index files on the slave server, and then move the generated index files to the data/index folder on the main server. It is precisely because of this index update method that I need to notify the web side in real-time that the index has been updated. But it seems that there is no such API for such operations

**Reply by vladak** [reply]
There is: [projects/{project}/indexed](https://opengrok.docs.apiary.io/#reference/0/project-metadata-management/marks-project-as-indexed) - this will refresh the index searchers. This is what the indexer itself uses when done indexing a project (assuming it was run with -U).

Does the incremental reindex take so long that you need to move the reindex process off the main machine ? Normally incremental reindex should be very fast (even more so if the repositories in given project are all Git or Mercurial - assuming recent OpenGrok version), unless there are many changes to the project in question.

**Reply by qintian-qe** [reply]
Yes, incremental indexing is already fast, but the code volume is too large, and the total time required for code updates combined with index updates cannot meet the requirement of daily code updates

**Reply by vladak** [reply]
How large it "too large" ? Also, have you tried using the [per project workflow](https://github.com/oracle/opengrok/wiki/Per-project-management-and-workflow) ?

**Reply by qintian-qe** [reply]
Approximately 10T of code, and the file size is also very large. I saw in the document that (due to scanning the entire project directory tree for changed files) even though the incremental changes are small

**Reply by vladak** [reply]
> Approximately 10T of code, and the file size is also very large. I saw in the document that (due to scanning the entire project directory tree for changed files) even though the incremental changes are small

The directory scan is avoided in certain conditions (described above).
